pandas/core/indexes/numeric.py

```diff
@@ -301,6 +332,26 @@ def _assert_safe_casting(cls, data, subarr):
                 raise TypeError('Unsafe NumPy casting, you must '
                                 'explicitly cast')
 
+    def _is_compatible_with_other(self, other):
+        return (
+            super()._is_compatible_with_other(other)
+            or all(isinstance(type(obj), (ABCUInt64Index,
+                                          ABCFloat64Index))
+                   for obj in [self, other])
+        )
+
+    def _union(self, other, sort):
+        needs_cast = (
+            (is_integer_dtype(self.dtype) and is_float_dtype(other.dtype)) or
+            (is_integer_dtype(other.dtype) and is_float_dtype(self.dtype))
+        )
+        if needs_cast:
```

Comment:
maybe makes sense to put this in a private method (the needs_cast logic) as its repeated 3 X

Reply:
Whoops, these can all use NumericIndex._union. Forgot to remove from here.